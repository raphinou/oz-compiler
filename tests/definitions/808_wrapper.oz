% Secure data storage wrapper
local
   NewWrapper
   Wrap
   Unwrap
   Five
   Six
   R
in


proc {NewWrapper ?Wrap ?Unwrap}
   Key={NewName}
in
   fun {Wrap X} {Chunk.new w(Key:X)} end
   fun {Unwrap W}
      try W.Key
      catch _ then raise error(unwrap(W)) end end
   end
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

{NewWrapper Wrap Unwrap}
Five={Wrap 5}
Six={Wrap 6}
R={Wrap rec(1 2 3)}
{Show Five}
{Show {Unwrap Five}}
{Show Six}
{Show {Unwrap Six}}
{Show R}
   {Show {Unwrap R}}
end
