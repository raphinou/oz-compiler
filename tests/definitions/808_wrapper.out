<Chunk>
5
<Chunk>
6
<Chunk>
rec(1 2 3)
